verification/spec/relayer/environment.tla

```diff
@@ -0,0 +1,275 @@
+ValidClientID(chainID, clientID) ==
+    /\ GetChainIDs(clientID).src = chainID
+    /\ GetChainIDs(clientID).dst \in ChainIDs 
+    
+(***************
+Chain update operators
+***************)    
+
+LightClientUpdate(chainID, chain, datagrams) == 
+    \* update chain with "CreateClient" datagrams
+    LET createClients == {dgr \in datagrams : dgr.type = "CreateClient"} IN
+    LET createClientIDs == {clientID \in ClientIDs : 
+            /\ \E dgr \in createClients : dgr.clientID = clientID
+            /\ ValidClientID(chainID, clientID)} IN
+    LET createClientHeights(clientID) == 
+            {h \in Heights : 
+                \E dgr \in createClients : 
+                    dgr.clientID = clientID /\ dgr.height = h} IN  
+    
+    \* new chain record with clients created
+    LET clientsCreated == [
+            height |-> chain.height,
+            clientHeights |-> 
+                [cID \in ChainIDs |-> IF GetClientID(chainID, cID) \in createClientIDs
+                                      THEN Max(createClientHeights(GetClientID(chainID, cID)))
+                                      ELSE chain.clientHeights[cID] 
+                ]
+         ] IN
+     
+    \* update chain with "ClientUpdate" datagrams
+    LET updateClients == {dgr \in datagrams : dgr.type = "ClientUpdate"} IN
+    LET updateClientIDs == {clientID \in ClientIDs : 
+            /\ \E dgr \in updateClients : dgr.clientID = clientID
+            /\ ValidClientID(chainID, clientID)} IN
+    LET updateClientHeights(clientID) == 
+            {h \in Heights : \E dgr \in updateClients : 
+                dgr.clientID = clientID /\ dgr.height = h} IN  
+    
+    \* new chain record with clients updated
+    LET clientsUpdated == [
+            height |-> clientsCreated.height,
+            clientHeights |-> 
+                [cID \in ChainIDs |-> IF GetClientID(chainID, cID) \in updateClientIDs
+                                      THEN Max(updateClientHeights(GetClientID(chainID, cID)))
+                                      ELSE clientsCreated.clientHeights[cID] 
+                ]
+         ] IN
+    
+    clientsUpdated
+ 
+\* Update chainID with the received datagrams
+\* Currently, only supporting ICS2: Client updates
+\* TODO: Support the remaining datagrams
+UpdateChain(chainID, datagrams) == 
+    LET chain == chains[chainID] IN
+    LET lightClientsUpdated == LightClientUpdate(chainID, chain, datagrams) IN 
+    
+    lightClientsUpdated
+    
+(***************
+Chain actions
+***************)       
+
+\* Advance the height of the chain until MaxHeight is reached
+AdvanceChain ==
+    \E chainID \in ChainIDs :
+        /\ chains[chainID].height < MaxHeight
+        /\ chains' = [chains EXCEPT 
+                        ![chainID].height = chains[chainID].height + 1                                            
+                     ]
+        /\ UNCHANGED pendingDatagrams
+
+\* Update the height of the light client on chainID for chainID 
+UpdateOwnClient ==
```

Comment:
Is this needed if we let the relayer update with src=dest? Or initialize chains not to create its own clients?
I will have to confirm if we support this currently but in principle we allow loopback connections and they would require a client for a chain A to be created on chain A. But this would also be done by the relayer.

Reply:
We can probably remove this for sake of simplicity (at least initially).

Reply:
I was looking at the ICS, in ICS 018 in the specification of `pendingDatagrams`, the relayer is only creating datagrams when src != dst. But in principle, we could allow relayer updates when src = dst. 

Are loopback connections those where src = dst? I.e., is it allowed that chain A creates a connection with itself? I'm modeling the connection datagrams now and I wasn't sure if this is allowed.